src/selectors/pipeline.js

```diff
@@ -0,0 +1,50 @@
+import { createSelector } from 'reselect';
+import { arrayToObject } from '../utils';
+
+const getNodeIDs = state => state.node.ids;
+const getNodePipelines = state => state.node.pipelines;
+const getActivePipeline = state => state.pipeline.active;
+const getTagIDs = state => state.tag.ids;
+const getNodeTags = state => state.node.tags;
+
+/**
+ * Calculate whether nodes should be disabled based on their tags
+ */
+export const getNodeDisabledPipeline = createSelector(
+  [getNodeIDs, getNodePipelines, getActivePipeline],
+  (nodeIDs, nodePipelines, activePipeline) =>
+    arrayToObject(nodeIDs, nodeID => {
+      if (!activePipeline) {
+        return false;
+      }
+      return !nodePipelines[nodeID][activePipeline];
+    })
+);
+
+/**
+ * Get a list of just the IDs for the active pipeline
+ */
+export const getPipelineNodeIDs = createSelector(
+  [getNodeIDs, getNodeDisabledPipeline],
+  (nodeIDs, nodeDisabledPipeline) =>
+    nodeIDs.filter(nodeID => !nodeDisabledPipeline[nodeID])
+);
+
+/**
+ * Get IDs of tags used in the current pipeline
+ */
+export const getPipelineTagIDs = createSelector(
+  [getTagIDs, getPipelineNodeIDs, getNodeTags],
+  (tagIDs, nodeIDs, nodeTags) => {
+    const visibleTags = {};
+    for (let tagID of tagIDs) {
```

Comment:
This function appears to be something like `O(ts·ns·k)` i.e. `O(n^3)` worst case, is that right? I have a feeling this might be refactored to be linear instead?

Reply:
I'd welcome your suggestions. This is actually what I came up with after refactoring my initial attempt, which was something like this:
```es6

    const visibleTags = {};
    for (let nodeID of nodeIDs) {
      for (let tagID of nodeTags[nodeID]) {
        if (!visibleTags[tagID]) {
          visibleTags[tagID] = true;
        }
      }
    }
```
which would keep going through all the nodes even after checking all the tags. I think the one I came up with is better, and it seems to be faster in testing (seems to usually take about half as many operations).

Reply:
Yeah I realise that it's fast enough in practice for now, but I think it's a good idea to avoid this anyway if there's an equally simple linear version.

Would this do the trick?

```js
export const getPipelineTagIDs = createSelector(
  [getPipelineNodeIDs, getNodeTags],
  (nodeIDs, nodeTags) => {
    const tagIds = {};

    for (const nodeID of nodeIDs) {
      for (const tagID of nodeTags[nodeID]) {
        tagIds[tagID] = true;
      }
    }

    return Object.keys(tagIds);
  }
);
```

Reply:
yep that's pretty much Option B, but without the filter (which I was undecided on). Honestly it depends on the data shape and how many tags there are vs the number of nodes. But yeah I might as well use that one. done.